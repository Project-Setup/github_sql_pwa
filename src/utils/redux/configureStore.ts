import {
  configureStore,
  combineReducers,
  ConfigureStoreOptions,
  ReducersMapObject,
  Reducer,
  Store,
  Action,
  Middleware,
  AnyAction,
} from '@reduxjs/toolkit';
import objectAssign from '../common/objectAssign';

interface StoreReducerEnhanced {
  reducer: ReducersMapObject;
  addReducer: <S = any, A extends Action<any> = AnyAction>(
    key: string,
    addedReducer: Reducer<S, A>
  ) => void;
  removeReducers: (keys: string[]) => void;
  injectReducers: <S = any, A extends Action<any> = AnyAction>(
    reducers: ReducersMapObject<S, A>
  ) => void;
  substituteReducers: <S = any, A extends Action<any> = AnyAction>(
    reducers: ReducersMapObject<S, A>
  ) => void;
}

export interface ReducerEnhancedStore<
  S = any,
  A extends Action<any> = AnyAction
> extends Store<S, A>, StoreReducerEnhanced {}

export interface ModifiedConfigureStoreOptions<
  S = any,
  A extends Action<any> = AnyAction,
  M extends ReadonlyArray<Middleware<{}, S>> = ReadonlyArray<Middleware<{}, S>>
> extends ConfigureStoreOptions<S, A, M> {
  reducer: ReducersMapObject<S, A>;
}

const modifiedConfigureStore = <S = any, A extends Action<any> = AnyAction>({
  reducer,
  ...restConfig
}: ModifiedConfigureStoreOptions<S, A>) => {
  let keysToRemove: string[] = [];

  const createReducer = <
    S extends object = {},
    A extends Action<any> = AnyAction
  >(
    reducerObj: ReducersMapObject<S, A>
  ): Reducer<S, A> => (state = {} as S, action) => {
    let updatedState = state;
    if (keysToRemove.length > 0) {
      updatedState = objectAssign(([k]) => !keysToRemove.includes(k))(
        {},
        state
      );
      keysToRemove = [];
    }
    return combineReducers<S, A>(reducerObj)(updatedState, action);
  };

  const store: ReducerEnhancedStore<S, A> = Object.assign(
    configureStore({ reducer, ...restConfig }),
    {
      reducer,

      addReducer: (key, addedReducer) => {
        if (!key || store.reducer[key]) {
          return;
        }
        store.reducer[key] = addedReducer as Reducer<any, AnyAction>;
        store.replaceReducer(createReducer(store.reducer));
      },

      removeReducers: (keys) => {
        keys.forEach((key) => {
          if (!key || !store.reducer[key]) {
            return;
          }
          keysToRemove.push(key);
        });
        store.reducer = objectAssign(([k]) => !keysToRemove.includes(k))(
          {},
          store.reducer
        );
        store.replaceReducer(createReducer(store.reducer));
      },

      injectReducers: (reducers) => {
        objectAssign()(store.reducer, reducers);
        store.replaceReducer(createReducer(store.reducer));
      },

      substituteReducers: (reducers) => {
        keysToRemove.push(
          ...Object.keys(store.reducer).filter((k) => !(k in reducers))
        );
        store.reducer = objectAssign()({}, reducers);
        store.replaceReducer(createReducer(store.reducer));
      },
    } as StoreReducerEnhanced
  );

  return store;
};

export default modifiedConfigureStore;
