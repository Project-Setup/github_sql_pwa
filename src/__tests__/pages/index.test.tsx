/* eslint-disable react/jsx-props-no-spreading */
import React, { ReactChild } from 'react';
import { mount, ReactWrapper } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore, {
  ModifiedConfigureStoreOptions,
} from 'utils/redux/configureStore';
import IndexPage from 'pages/index';
import { act } from 'react-dom/test-utils';

const mountWithStore = (children: ReactChild) => ({
  reducer = {},
  devTools = false,
  middleware,
  preloadedState,
  enhancers,
}: ModifiedConfigureStoreOptions = {}) => {
  const store = configureStore({
    reducer,
    devTools,
    middleware,
    preloadedState,
    enhancers,
  });
  return mount(<Provider store={store}>{children}</Provider>);
};

describe('IndexPage', () => {
  const testProp = {
    test: 'test',
  };

  let TestIndexPageWrapper: ReactWrapper;

  beforeAll(async () => {
    await act(async () => {
      TestIndexPageWrapper = mountWithStore(<IndexPage {...testProp} />)();
    });
  });

  it('should render without throwing an error', () => {
    expect(TestIndexPageWrapper).toMatchSnapshot();
  });

  it('should render appProps correctly', () => {
    expect(TestIndexPageWrapper.find('p').at(0).text()).toContain(
      JSON.stringify(testProp)
    );
  });

  it('should render page process env correctly', () => {
    expect(TestIndexPageWrapper.find('p').at(1).text()).toContain(
      process.env.NEXT_PUBLIC_TEST_PAGE_VAR
    );
  });
});
