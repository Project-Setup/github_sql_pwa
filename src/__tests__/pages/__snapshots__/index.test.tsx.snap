// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndexPage should render without throwing an error 1`] = `
<Provider
  store={
    Object {
      "addReducer": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "injectReducers": [Function],
      "reducer": Object {
        "count": [Function],
      },
      "removeReducers": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      "substituteReducers": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <IndexPage
    test="test"
  >
    <main>
      <ManifestHead
        hrefCanonical="/"
        title="github_pwa static-path"
      >
        <Head>
          <Component
            handleStateChange={null}
            inAmpMode={false}
            reduceComponentsToState={[Function]}
          />
        </Head>
      </ManifestHead>
      <p>
        Props from _app.tsx: {"test":"test"}
      </p>
      <p>
        page process env: TEST_TEST_PAGE_PROP
      </p>
      <CountDivWrap>
        <DynamicStoreWrap
          callbackOnMount={[Function]}
          callbackOnUnmount={[Function]}
        >
          <DynamicStore
            callbackOnMount={[Function]}
            callbackOnUnmount={[Function]}
            store={
              Object {
                "addReducer": [Function],
                "dispatch": [Function],
                "getState": [Function],
                "injectReducers": [Function],
                "reducer": Object {
                  "count": [Function],
                },
                "removeReducers": [Function],
                "replaceReducer": [Function],
                "subscribe": [Function],
                "substituteReducers": [Function],
                Symbol(observable): [Function],
              }
            }
          />
        </DynamicStoreWrap>
      </CountDivWrap>
      <PrefixedLink
        href="/index/1"
      >
        <Link
          as="/index/1"
          href="/index/1"
        >
          <a
            href="/index/1"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            index/1
          </a>
        </Link>
      </PrefixedLink>
    </main>
  </IndexPage>
</Provider>
`;
